base_serializer.go

```diff
@@ -24,48 +30,155 @@ type baseSerializer struct {
-	panic("implement me")
+	baseSerializer.hashFunction.Update([]byte{EndMarkerByte})
 }
 
-func (baseSerializer *baseSerializer) handleAnnotationsBegin(ionValue interface{}, isContainer bool) {
-	panic("implement me")
+func (baseSerializer *baseSerializer) handleAnnotationsBegin(ionValue hashValue, isContainer bool) error {
+	annotations := ionValue.getAnnotations()
+	if len(annotations) > 0 {
+		baseSerializer.beginMarker()
+		baseSerializer.update([]byte{TqValue})
+		for _, annotation := range annotations {
+			err := baseSerializer.writeSymbol(annotation)
+			if err != nil {
+				return err
+			}
+		}
+
+		if isContainer {
+			baseSerializer.hasContainerAnnotation = true
+		}
+	}
+
+	return nil
 }
 
-func (baseSerializer *baseSerializer) handleAnnotationsEnd(ionValue interface{}, isContainer bool) {
-	panic("implement me")
+func (baseSerializer *baseSerializer) handleAnnotationsEnd(ionValue hashValue, isContainer bool) {
+	if (ionValue != nil && len(ionValue.getAnnotations()) > 0) ||
+		(isContainer && baseSerializer.hasContainerAnnotation) {
+		baseSerializer.endMarker()
+
+		if isContainer {
+			baseSerializer.hasContainerAnnotation = false
+		}
+	}
 }
 
-func (baseSerializer *baseSerializer) writeSymbol(token string) {
-	panic("implement me")
+func (baseSerializer *baseSerializer) writeSymbol(token string) error {
+	baseSerializer.beginMarker()
+
+	// SymbolTokens are not available right now.
+	/*var sid int
+	if token == "" {
+		sid = 0
+	} else {
+		sid = ion.SymbolToken.UnknownSid
+	}
+
+	symbolToken := &ion.SymbolToken{token, sid}
+	scalarBytes, err := baseSerializer.getBytes(ion.SymbolType, symbolToken, false);
+	if err != nil {
+		return nil
+	}
+
+	tq, representation := baseSerializer.scalarOrNullSplitParts(ion.SymbolType, symbolToken, false, scalarBytes)
+
+	baseSerializer.update([]byte{tq})
+	if len(representation) > 0 {
+		baseSerializer.update(escape(representation))
+	}*/
+
+	baseSerializer.endMarker()
+
+	return nil
 }
 
-func (baseSerializer *baseSerializer) getBytes(ionType ion.Type, ionValue interface{}, isNull bool) []byte {
-	panic("implement me")
+func (baseSerializer *baseSerializer) getBytes(ionType ion.Type, ionValue interface{}, isNull bool) ([]byte, error) {
+	if isNull {
+		typeCode := byte(ionType)
+		return []byte{(typeCode << 4) | 0x0F}, nil
+	} else if ionType == ion.FloatType && ionValue == 0 && int64(ionValue.(float64)) >= 0 {
+		// value is 0.0, not -0.0
+		return []byte{0x40}, nil
+	} else {
+		buf := bytes.Buffer{}
+		writer := NewHashWriter(ion.NewBinaryWriter(&buf), NewCryptoHasherProvider(SHA256))
+
+		err := serializers(ionType, ionValue, writer)
+		if err != nil {
+			return nil, err
+		}
+
+		err = writer.Finish()
+		if err != nil {
+			return nil, err
+		}
+
+		return buf.Bytes()[4:], nil
+	}
 }
 
-func (baseSerializer *baseSerializer) getLengthLength(bytes []byte) int {
-	panic("implement me")
+func (baseSerializer *baseSerializer) getLengthLength(bytes []byte) (int, error) {
```

Comment:
maybe `lengthFieldLength` is a better name?

Reply:
Renamed getLengthLength() to getLengthFieldLength()